spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/bigquant/SpatialConvolution.scala

```diff
@@ -0,0 +1,382 @@
+      val groupWeight = weightFP32.select(1, i)
+      val bufferOffset = 0
+      val buffer = new Array[Byte](groupWeight.nElement())
+      ev.getType() match {
+        case FloatType =>
+          Quant.quantize(groupWeight.toTensor[Float], buffer, bufferOffset)
+        case _ => throw new UnsupportedOperationException(s"Only support Float for quantized model")
+      }
+
+      weight(i - 1).setStorage(buffer)
+    }
+
+    init()
+
+    this
+  }
+
+  @throws(classOf[IOException])
+  private def writeObject(out: ObjectOutputStream): Unit = {
+    out.defaultWriteObject()
+
+    out.writeObject(weight)
+  }
+
+  @throws(classOf[IOException])
+  private def readObject(in: ObjectInputStream): Unit = {
+    in.defaultReadObject()
+
+    weight = in.readObject().asInstanceOf[Array[QuantTensor[T]]]
+    if (weight(0).getStorage != null && weight(0).getStorageInJni == 0L) {
+      init()
+    }
+  }
+
+  private def init(): this.type = {
+    for (i <- 1 to nGroup) {
+      if (weight(i - 1).getStorageInJni == 0L) {
+        val byteArrayOfWeight = weight(i - 1).getStorage
+        weight(i - 1).setStorageInJni(
+          BigQuant.ConvKernelDescInit(nOutputPlane / nGroup, nInputPlane / nGroup,
+            kernelH, kernelW))
+        ev.getType() match {
+          case FloatType =>
+            val start = (i - 1) * nOutputPlane / nGroup
+            val end = i * nOutputPlane / nGroup
+            val minArray = min.asInstanceOf[Array[Float]].slice(start, end)
+            val maxArray = max.asInstanceOf[Array[Float]].slice(start, end)
+            val byteOffset = 0
+
+            BigQuant.ConvKernelLoadFromModel(weight(i - 1).getStorageInJni,
+              byteArrayOfWeight, byteOffset,
+              minArray, maxArray, nOutputPlane / nGroup, nInputPlane / nGroup,
+              kernelH, kernelW, WEIGHT_THRESHOLD, BigQuant.NCHW)
+          case _ =>
+            throw new UnsupportedOperationException(s"Only support Float for quantized model")
+        }
+      }
+    }
+
+    _init = true
+
+    this
+  }
+
+  private def checkAndInit(): Unit = {
+    if (!_init) {
+      init()
+    }
+  }
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim() == 3 || input.dim() == 4,
+      "bigquant.SpatialConvolution: " + ErrorInfo.constrainInputAs3DOrBatch)
+    require(input.isContiguous())
+
+    checkAndInit()
+
+    // compute attributes of input and output
+    val (batchSize, inputHeight, inputWidth) = if (input.dim() == 3) {
+      (1, input.size(2), input.size(3))
+    } else {
+      (input.size(1), input.size(3), input.size(4))
+    }
+
+    val outputHeight = outputSize(inputHeight, padH, kernelH, strideH)
+    val outputWidth = outputSize(inputWidth, padW, kernelW, strideW)
+
+    output.resize(Array(batchSize, nOutputPlane, outputHeight, outputWidth))
+
+    if (!data.isInitialized) {
+      data.setStorageInJni(BigQuant.ConvDataDescInit(nInputPlane / nGroup, kernelH, kernelW,
```

Comment:
What if input height and width are changed after data is initialized?

Reply:
Currently, it only supports dynamic batch size, but not input height and input width. For this case, I have to reinitialize it. I'll add it.